src/uffd_handler/src/handler.rs

```diff
@@ -0,0 +1,432 @@
+    }
+
+    fn serve_page_fault(&mut self, addr: *mut u8) -> Result<()> {
+        let fault_page_addr = page_start_of_addr(addr)?;
+
+        // Get the state of the current faulting page.
+        for region in self.mem_regions.iter() {
+            match region.page_states.get(&fault_page_addr) {
+                // Our simple PF handler has a simple strategy:
+                // There exist 4 states in which a memory page can be in:
+                // 1. Uninitialized - page was never touched
+                // 2. FromFile - the page is populated with content from snapshotted memory file
+                // 3. Removed - MADV_DONTNEED was called due to balloon inflation
+                // 4. Anonymous - page was zeroed out -> this implies that more than one page fault
+                //    event was received. This can be a consequence of guest reclaiming back its
+                //    memory from the host (through balloon device)
+                Some(MemPageState::Uninitialized) | Some(MemPageState::FromFile) => {
+                    let (start, end) = self
+                        .uffd
+                        .populate_from_file(self.backing_buffer as u64, region)?;
+                    self.update_mem_state_mappings(start, end, &MemPageState::FromFile);
+                    return Ok(());
+                }
+                Some(MemPageState::Removed) | Some(MemPageState::Anonymous) => {
+                    let (start, end) = self.uffd.zero_out(fault_page_addr)?;
+                    self.update_mem_state_mappings(start, end, &MemPageState::Anonymous);
+                    return Ok(());
+                }
+                None => {}
+            }
+        }
+
+        Err(HandlerError::AddressNotInRegion)
+    }
+
+    fn handle_event(&mut self, event: Event) -> Result<()> {
+        // Expect to receive either a `PageFault` or `Removed` event
+        // (if the balloon device is enabled).
+        match event {
+            Event::Pagefault { addr, .. } => self.serve_page_fault(addr.cast::<u8>())?,
+            Event::Remove { start, end } => self.update_mem_state_mappings(
+                start.cast::<u8>() as u64,
+                end.cast::<u8>() as u64,
+                &MemPageState::Removed,
+            ),
+            event => return Err(HandlerError::UnexpectedEvent(event)),
+        }
+        Ok(())
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        let mut pollfd = libc::pollfd {
+            fd: self.uffd.as_raw_fd(),
+            events: libc::POLLIN,
+            revents: 0,
+        };
+
+        // Loop, handling incoming events on the userfaultfd file descriptor.
+        loop {
+            // SAFETY: Safe because fd, nfds and timeout are valid parameters.
+            let nready = unsafe { libc::poll(&mut pollfd, 1, POLL_TIMEOUT) };
+            // Poll has an infinite timeout, therefore in theory this case should never happen.
+            if nready == -1 {
+                unreachable!();
```

Comment:
seems like a signal could cause this: https://man7.org/linux/man-pages/man2/poll.2.html#ERRORS - maybe better to just `continue` instead of panicking?

Reply:
I think here it would be better to handle each case separately, for example for these error types we should terminate the process (retrying will most likely not solve the issue):
```
       EFAULT fds points outside the process's accessible address space.
              The array given as argument was not contained in the
              calling program's address space.

       EINVAL The nfds value exceeds the RLIMIT_NOFILE value.

       ENOMEM Unable to allocate memory for kernel data structures.
```

not 100% decided on this one, need to dive a little deeper:
```
       EINTR  A signal occurred before any requested event; see
              [signal(7)](https://man7.org/linux/man-pages/man7/signal.7.html).
```

Reply:
I think here it would be better to handle each case separately, for example for these error types we should terminate the process (retrying will most likely not solve the issue):
```
       EFAULT fds points outside the process's accessible address space.
              The array given as argument was not contained in the
              calling program's address space.

       EINVAL The nfds value exceeds the RLIMIT_NOFILE value.

       ENOMEM Unable to allocate memory for kernel data structures.
```

not 100% decided on this one, need to dive a little deeper:
```
       EINTR  A signal occurred before any requested event; see
              [signal(7)](https://man7.org/linux/man-pages/man7/signal.7.html).
```

Reply:
I updated the code to exit in case of `EFAULT`, `EINVAL` and `ENOMEM`, because these are not recoverable and we might end up looping endlessly. For  `EINTR`, the handler retries the operation, because the signal triggered before the event can be harmless. In case it is not, `poll` should fail after the retry with one of the unrecoverable errors.